### PR TITLE
Blog introducing PodTopologySpread

### PR DESCRIPTION
**Pull request by Huang-Wei**
A blog introducing PodTopologySpread. Reviewable on https://deploy-preview-20602--kubernetes-io-master-staging.netlify.app/blog/2020/04/introducing-podtopologyspread/.

### CODE REVIEW

**Comment by Huang-Wei** [comment]
cc/ @alculquicondor

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+title: "Introducing PodTopologySpread"
+date: 2020-04-27
+slug: introducing-podtopologyspread
+url: /blog/2020/04/Introducing-PodTopologySpread
+---
+
+**Author:** Wei Huang (IBM), Aldo Culquicondor (Google)
+
+Managing Pods distribution across a cluster is hard. The well-known Kubernetes
+features, PodAffinity and PodAntiAffinity, allow some control of Pod placement
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:11]: This is more commonly written as something like:
```suggestion
features for Pod affinity and anti-affinity allow some control of Pod placement
```

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+in different topologies. However, these features only resolve part of Pods
+distribution use cases: either place unlimited Pods to a single topology, or
+disallow two Pods to co-locate in the same topology. In between these two
+extreme cases, there is a common need to distribute the Pods evenly across the
+topologies, so as to achieve better cluster utilization and high availability of
+applications.
+
+The PodTologySpread scheduling plugin (originally proposed as EvenPodsSpread)
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:19]:
```suggestion
The PodToplogySpread scheduling plugin (originally proposed as EvenPodsSpread)
```

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    whenUnsatisfiable: <string>
+    labelSelector: <object>
+```
+
+As this API is embedded in Pod's spec, you can use this feature in all the
+high-level workload APIs, such as Deployment, DaemonSet, StatefulSet, etc.
+
+Let's see an example of a cluster to understand this API.
+
+![](/images/blog/2020-04-27-introducing-podtopologyspread/api.png)
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:40]: For accessibility, add an alt attribute.

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    hard constraint.
+  - `ScheduleAnyway` tells the scheduler to still schedule it while prioritizing
+    Nodes that reduce the skew. It's a soft constraint.
+
+## Advanced usage
+
+As the feature name "PodTopologySpread" implies, the basic usage of this feature
+is to run your workload with an absolute even manner (maxSkew=1), or relatively
+even manner (maxSkew>=2). See the [official
+document](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/)
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:70]:
```suggestion
documentation](/docs/concepts/workloads/pods/pod-topology-spread-constraints/)
```
?

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+search all Nodes and group them by "topologyKey". Sometimes this may not be the
+ideal case. For instance, suppose there is a cluster with Nodes tagged with
+"env=prod", "env=staging" and "env=qa", and now you want to evenly place Pods to
+the "qa" environment across zones, is it possible?
+
+The answer is yes. You can leverage the NodeSelector or NodeAffinity API spec.
+Under the hood, the PodTopologySpread feature will **honor** that and calculate
+the spread constraints among the nodes that satisfy the selectors.
+
+![](/images/blog/2020-04-27-introducing-podtopologyspread/advanced-usage-1.png)
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:89]: For accessibility, add an alt attribute.

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+As illustrated above, you can specify "spec.affinity.nodeAffinity" to limit the
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:91]: nit:
```suggestion
As illustrated above, you can specify `spec.affinity.nodeAffinity` to limit the
```

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+TopologySpreadConstraint is calculated independently, and the result sets will
+be merged to generate the eventual result set - i.e., suitable Nodes.
+
+In the following example, we want to schedule a Pod to a cluster with 2
+requirements at the same time:
+
+- place the Pod evenly with Pods across zones
+- place the Pod evenly with Pods across nodes
+
+![](/images/blog/2020-04-27-introducing-podtopologyspread/advanced-usage-2.png)
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:109]: For better accessibility, add an alt attribute.

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+authors need to be aware of the underlying topology of the cluster, and then
+specify proper `topologySpreadConstraints` in the Pod spec for every workload.
+While the Pod-level API gives the most flexibility it is also possible to
+specify cluster-level defaults.
+
+The default PodTopologySpread constraints allow you to specify spreading for all
+the workloads in the cluster, tailored for its topology. The constraints can be
+specified by an operator/admin as PodTopologySpread plugin arguments in the
+[scheduling profile configuration
+API](https://kubernetes.io/docs/reference/scheduling/profiles/) when starting
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:150]:
```suggestion
API](/docs/reference/scheduling/profiles/) when starting
```
?

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    args:
+      defaultConstraints:
+      - maxSkew: 1
+        topologyKey: example.com/rack
+        whenUnsatisfiable: ScheduleAnyway
+```
+
+When configuring default constraints, label selectors must be left empty.
+kube-scheduler will deduce the label selectors from the membership of the pod to
+services, replication controllers, replica sets or stateful sets. Pods can
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:170]: nit:
```suggestion
Services, ReplicaSets, or StatefulSets. Pods can
```
?

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+also combine using "hard" constraints and "soft" constraints to adhere to more
+diverse cluster situations.
+
+{{< note >}}
+If two TopologySpreadConstraints are being applied for the same {topologyKey,
+whenUnsatisfiable} tuple, the Pod creation will be blocked returning a
+validation error.
+{{< /note >}}
+
+## DefaultPodTopologySpread
```
Comment (sftim) [inline content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:138]:
```suggestion
## PodTopologySpread defaults
```
?

**Comment by Huang-Wei** [comment]
Thanks @sftim for the review! Your comments have been addressed.

**Comment by Huang-Wei** [comment]
Kindly ping @kbarnard10 @mrbobbytables

File: content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+A new field `topologySpreadConstraints` is introduced in the Pod's spec API:
+
+```yaml
+spec:
+  topologySpreadConstraints:
+  - maxSkew: <integer>
+    topologyKey: <string>
+    whenUnsatisfiable: <string>
+    labelSelector: <object>
+```
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:33]: Thanks Bob! I just removed the `yaml` designator.

BTW: before getting it merged, is it required to update the file name to reflect the latest date?

[the remark this answers is not in the document]
Comment (mrbobbytables) [reply content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:33]: Yes, but that's usually handled by the blog team before final publish -- reason being we never really know the exact date until everyone has signed off. It's kind of a pain =/ but is what it is.

[the remark this answers is not in the document]
Comment (sftim) [reply content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:33]: Ah, yes, it's not valid YAML. I agree, it makes sense to turn off highlighting.

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply content/en/blog/_posts/2020-04-27-introducing-podtopologyspread.md:33]: Got it. Thanks Bob and Tim!

**Comment by onlydole** [comment]
Thank you, @Huang-Wei for your contribution! 

/lgtm

**Comment by mrbobbytables** [comment]
/approve
/hold
will release the hold tomorrow during the day for it to go live 👍

**Comment by Huang-Wei** [comment]
> will release the hold tomorrow during the day for it to go live 👍

@mrbobbytables I assume you will cancel /hold today, right? Or, you want me to cancel it?

**Comment by mrbobbytables** [comment]
@Huang-Wei either of us could at this point, was just waiting till after lunch XD
With that though....
/hold cancel

**Pull request merged**